### PR TITLE
Allow a pathlib.PurePath object as a input to open_filename

### LINKED ISSUE #491: open_filename doesn't accept a pathlib object
**Bug report**

If the `filename` argument is a `pathlib` object, it is mistakenly recognized as a file handler.
I think `open_filename` should accept a `pathlib` object.
https://github.com/pdfminer/pdfminer.six/blob/0b44f7771462363528c109f263276eb254c4fcd0/pdfminer/utils.py#L14-L25

### PR DESCRIPTION
**Pull request by estshorter**
**Pull request**

This resolves #491 

**How Has This Been Tested?**

Tests added in this PR will verify my changes.

**Checklist**

- [x] I have added tests that prove my fix is effective or that my feature 
  works
- [ ] I have added docstrings to newly created methods and classes
- [x] I have optimized the code at least one time after creating the initial 
  version
- [x] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [x] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [x] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: pdfminer/utils.py [hunk 1]
```diff
@@ -17,6 +18,8 @@ class open_filename(object):
     (just like `open`), but does nothing for file-like objects.
     """
     def __init__(self, filename, *args, **kwargs):
+        if isinstance(filename, pathlib.PurePath):
+            filename = str(filename)
         if isinstance(filename, str):
             self.file_handler = open(filename, *args, **kwargs)
             self.closing = True
```
Comment (pietermarsman) [inline pdfminer/utils.py:25]: While we are changing this code, could you also add a conditional to the file_handler clause? E.g. use `elif` and raise an exception in the `else` clause that this type of input is not supported.

File: tests/test_highlevel_extracttext.py [hunk 2]
```diff
@@ -60,6 +69,26 @@ def test_simple4_with_string(self):
...
+    def test_simple3_with_pathlib(self):
+        test_file = "simple3.pdf"
+        s = run_with_pathlib(test_file)
+        self.assertEqual(s, test_strings[test_file])
+
+    def test_simple4_with_pathlib(self):
+        test_file = "simple4.pdf"
+        s = run_with_pathlib(test_file)
+        self.assertEqual(s, test_strings[test_file])
+
```
Comment (pietermarsman) [inline tests/test_highlevel_extracttext.py:91]: No need to add these end-to-end tests for this feature. It's already properly tested by the similiar tests. Could you add tests specifically for `absolute_sample_path` instead?

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,11 @@ All notable changes in pdfminer.six will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## [Unreleased]
+
+### Changed
+- `open_filename` accepts a `pathlib.PurePath` object
```
Comment (pietermarsman) [inline CHANGELOG.md:9]: Add link to the PR at the end

**Review by pietermarsman — CHANGES_REQUESTED** [review]

File: tests/test_highlevel_extracttext.py (hunk 2 above)
Comment (estshorter) [reply tests/test_highlevel_extracttext.py:91]: Tests for `absolute_sample_path`? Would you give a code example? I'm a little bit confused.

File: CHANGELOG.md (hunk 3 above)
Comment (estshorter) [reply CHANGELOG.md]: Fixed in 20a51a4

File: pdfminer/utils.py (hunk 1 above)
Comment (estshorter) [reply pdfminer/utils.py:26]: Fixed in 023a0e3

File: tests/test_highlevel_extracttext.py (hunk 2 above)
Comment (estshorter) [reply tests/test_highlevel_extracttext.py:91]: Do you mean tests for `open_filename` like 4ece1faa557838786c0e6f54a75199bfaa9c5336? (not pushed to develop yet)

Comment (pietermarsman) [reply tests/test_highlevel_extracttext.py:91]: Yes. Test the happy flow.

File: pdfminer/utils.py [hunk 4]
```diff
@@ -17,12 +19,16 @@ class open_filename(object):
     (just like `open`), but does nothing for file-like objects.
```
Comment (pietermarsman) [inline pdfminer/utils.py:19]: Also add `pathlib.PurePath` to the docstring.

**Review by pietermarsman — CHANGES_REQUESTED** [review]
Almost done :)

File: tests/test_highlevel_extracttext.py (hunk 2 above)
Comment (estshorter) [reply tests/test_highlevel_extracttext.py]: Fixed in 575078b.

File: pdfminer/utils.py (hunk 4 above)
Comment (estshorter) [reply pdfminer/utils.py:20]: Addressed in db64da2, but please correct it if needed as I'm not a native English speaker.

**Review by pietermarsman — APPROVED** [review]
@estshorter Thanks for all the work!

**Pull request merged**